doc(1.7.0): support GKE Container-Optimized OS (#884)
longhorn/website: content/docs/1.7.0/advanced-resources/os-distro-specific/container-optimized-os-support.md

---

> title:  Container-Optimized OS Support

Reviewer: ```suggestion
title:  Container-Optimized OS (COS) Support
```

weight: 5
---

## Requirements

> **Note:**
> Longhorn currently supports Container-Optimized OS only when used as the base image for Google Kubernetes Engine (GKE), which includes a pre-configured Kubernetes environment. The following information may not apply to manually created Kubernetes environments, including Kubernetes provisioned with other orchestrators.

> The Container-Optimized OS does not include a package manager and does not allow non-containerized applications to run. Additionally, its root filesystem is mounted as read-only, which poses a challenge for IO operations.

Reviewer: Add a link for the GKE COS doc

```suggestion
The [Container-Optimized OS](https://cloud.google.com/container-optimized-os/docs) does not include a package manager and does not allow non-containerized applications to run. Additionally, its root filesystem is mounted as read-only, which poses a challenge for IO operations.
```

> In GKE, Kubernetes tackes these constraints by housing necessary dependencies in a chroot environment (`/home/kubernetes/containerized_mounter/rootfs`) and mounting directories within it, enabling the execution of required tasks.

Reviewer: ```suggestion
In GKE, Kubernetes tackles these constraints by housing necessary dependencies in a chroot environment (`/home/kubernetes/containerized_mounter/rootfs`) and mounting directories within it, enabling the execution of required tasks.
```

tackles?

> Longhorn provides a GKE COS node agent, which leverages GKE Kubernetes solutions to configure and run necessary dependencies. This agent is responsible for the following operations:

Reviewer: Can you provide more information or links to the official documents for the GKE COS node agent?

Reviewer: No. I'm a bit confused when reading `Longhorn provides a GKE COS node agent ...`.  Is "node agent" a tool, defined by GKE platform, used for installation or configuration of the environment in the GKE OS system? Is it provided by Longhorn? Is this term visible in the GKE OS documentation, or is it something we invented ourselves?

Author: It's a DaemonsSet Longhorn provides to the user for setting up dependencies and running the iSCSI daemon. Let me add the `DaemonSet` into the sentence to see if it is less confusing.

Reviewer: Thanks for your explanation. :)

> Longhorn provides a GKE COS node agent daemonset, which leverages GKE Kubernetes solutions to configure and run necessary dependencies. This agent is responsible for the following operations:

Reviewer: Should we add `GKE COS node agent` into the scope of the tooling cli?

Author: https://github.com/longhorn/longhorn/issues/7927

- Mounting the Longhorn data path.
- Loading the kernel module.

> - Installing and running the iSCSI daemon.

Reviewer: Are the NFS utility and encryption tool installed in the agent?

Author: The part regarding NFS is mentioned in the `Installing NFSv4 client` section, Regarding encryption I presume you are asking about the `cryptsetup` binary? Yes, it's already in the COS, Unlike Talos Linux, COS supports bash, so it comes with some core utilities pre-installed..

    ```

1. Check the agent pod's status.
    ```
    $ kubectl get pod -l app=longhorn-gke-cos-node
    NAME                                READY   STATUS    RESTARTS      AGE
    longhorn-gke-cos-node-agent-222w8   1/1     Running   1 (86m ago)   86m
    longhorn-gke-cos-node-agent-8r26h   1/1     Running   1 (86m ago)   86m
    longhorn-gke-cos-node-agent-nwhsw   1/1     Running   1 (86m ago)   86m
    ```

1. Check the installation result in the agent pod logs.
    ```
    Completed!
    Keep the container running for iscsi daemon
    ```

1. Verify that the dependent kernel module is loaded. You must run the command on the host.
    ```
    $ lsmod | grep -q iscsi_tcp && echo "The iSCSI module is loaded" || echo "The iSCSI module is NOT loaded"
    The iSCSI module is loaded
    ```

1. Verify that the iSCSI daemon is running. You must run the command on the host.
    ```
    $ ps aux | grep -q '[i]scsid' && echo "The iSCSI daemon is running" || echo "The iSCSI daemon is NOT running"
    The iSCSI daemon is running
    ```

1. Verify that the Longhorn data path (`/var/lib/longhorn`) is mounted on the host. You must run the command on the host.
    ```

>     $ findmnt --noheadings "/var/lib/longhorn"

Reviewer 2: Can we support multiple longhorn disks with different mount points?

Author: Good question, I don't see why we couldn't do that. Let me make the update to the YAML file.

    /var/lib/longhorn /dev/sda1[/var/lib/longhorn] ext4   rw,relatime,commit=30
    ```

## Limitations

> In COS clusters, Longhorn currently supports only V1 data volumes.

Reviewer 2: Are there any challenges to adding extra block devices for v2?

Author: I don't believe so at the moment, but need some time to investigate. Thinking about supporting V2 in phase 2, following a similar approach to Talos. What do you think?

Reviewer 2: Sounds good. Let's create a ticket for it.

Author: https://github.com/longhorn/longhorn/issues/8229

---
title:  Container-Optimized OS Support
weight: 5
---

## Requirements

> You must meet the following requirements before installing Longhorn on a Container-Optimized OS (COS) cluster.

Reviewer 3: Where are the requirements? We cannot keep this sentence if the requirements are not mentioned at all.

Author: The requirements refer to the following section about the dependencies. Do you have any alternative suggestions for how we can convey this better?

Author: I've moved part of the dependency section to the requirement section. Let me know what you think.

Reviewer 3: Got it. The changes that you made work for me.

> **Note:** Longhorn currently supports Container-Optimized OS only when used as the base image for Google Kubernetes Engine (GKE). This limitation is due to GKE comes with a pre-configured Kubernetes environment. For self-prepared Kubernetes environment or Kubernetes provisioned with other orchestrators, different strategies my be required as the provided method may be ineffective.

Reviewer 3: ```suggestion
> **Note:** 
> Longhorn currently supports Container-Optimized OS only when used as the base image for Google Kubernetes Engine (GKE), which includes a pre-configured Kubernetes environment. The following information may not apply to manually created Kubernetes environments, including Kubernetes provisioned with other orchestrators.
```

> ### Dependencies Configuration and Installation
> 
> The Container-Optimized OS does not include a package manager and doesn't support running non-containerized application. Additionally, its root filesystem is mounted as read-only, which poses a challenge for IO operations.
> 
> In GKE, Kubernetes tackes these constraints by housing necessary dependencies in a chroot environment (/home/kubernetes/containerized_mounter/rootfs) and mounting directories within it, enablind the execution of required tasks.
> 
> Longhorn provides a GKE COS node agent, leveraging GKE Kubernetes solutions to configure an run necessary dependencies. This agent is responsible for:
> 
> - Mounting the Longhorn data path.
> - Loading the Kernel module.
> - Installing and running the iSCSI daemon.

Reviewer 3: ```suggestion
### Dependencies, Installation, and Configuration

The Container-Optimized OS does not include a package manager and does not allow non-containerized applications to run. Additionally, its root filesystem is mounted as read-only, which poses a challenge for IO operations.

In GKE, Kubernetes tackes these constraints by housing necessary dependencies in a chroot environment (`/home/kubernetes/containerized_mounter/rootfs`) and mounting directories within it, enabling the execution of required tasks.

Longhorn provides a GKE COS node agent, which leverages GKE Kubernetes solutions to configure and run necessary dependencies. This agent is responsible for the following operations:

- Mounting the Longhorn data path
- Loading the kernel module
- Installing and running the iSCSI daemon
```

Author: > ### Dependencies, Installation, and Configuration
I am not sure if we should separate these words. It seems to imply three distinct tasks, but actually, they are all related to dependencies (data path, module, package..). Maybe we could consider removing this title and integrating the contexts under the `Requirement` section instead. What do you think?

Reviewer 3: Oh, you meant "Installation and Configuration of Dependencies". The new section heading works.

> To install the Longhorn GKE COS node agent, execute the following command:

Reviewer 3: ```suggestion
1. Install the Longhorn GKE COS node agent.
```

```
kubectl apply -f https://raw.githubusercontent.com/longhorn/longhorn/v{{< current-version >}}/deploy/prerequisite/longhorn-gke-cos-node-agent.yaml
```

> After the installation, verify the agent pod's status using:

Reviewer 3: ```suggestion
1. Check the agent pod's status.
```

```
$ kubectl get pod -l app=longhorn-gke-cos-node
NAME                                READY   STATUS    RESTARTS      AGE
longhorn-gke-cos-node-agent-222w8   1/1     Running   1 (86m ago)   86m
longhorn-gke-cos-node-agent-8r26h   1/1     Running   1 (86m ago)   86m
longhorn-gke-cos-node-agent-nwhsw   1/1     Running   1 (86m ago)   86m
```

> Check the installation result through the agent pod logs:

Reviewer 3: ```suggestion
1. Check the installation result in the agent pod logs.
```

```
Completed!
Keep the container running for iscsi daemon
```

> You can also check if the dependent Kernel module is loaded by running command on the host:

Reviewer 3: ```suggestion
1. Verify that the dependent kernel module is loaded. You must run the command on the host.
```

```
$ lsmod | grep -q iscsi_tcp && echo "The iSCSI module is loaded" || echo "The iSCSI module is NOT loaded"
The iSCSI module is loaded
```

> Ensure the iSCSI daemon is running:

Reviewer 3: ```suggestion
1. Verify that the iSCSI daemon is running.
```

```
$ ps aux | grep -q '[i]scsid' && echo "The iSCSI daemon is running" || echo "The iSCSI daemon is NOT running"
The iSCSI daemon is running
```

> Verify if the Longhorn data path (/var/lib/longhorn) is mounted on the host:

Reviewer 3: ```suggestion
1. Verify that the Longhorn data path (`/var/lib/longhorn`) is mounted on the host.
```

```
$ findmnt --noheadings "/var/lib/longhorn"
/var/lib/longhorn /dev/sda1[/var/lib/longhorn] ext4   rw,relatime,commit=30
```

## Limitations

> - Exclusive to v1 data volume: currently, within a COS cluster, Longhorn only supports v1 data volume. The v2 data volume isn't currently supported in this environment.

Reviewer 3: ```suggestion
In COS clusters, Longhorn currently supports only V1 data volumes.
```

> **Note:**
> Longhorn currently supports Container-Optimized OS only when used as the base image for Google Kubernetes Engine (GKE), which includes a pre-configured Kubernetes environment. The following information may not apply to manually created Kubernetes environments, including Kubernetes provisioned with other orchestrators.

The [Container-Optimized OS (COS)](https://cloud.google.com/container-optimized-os/docs) does not include a package manager and does not allow non-containerized applications to run. Additionally, its root filesystem is mounted as read-only, which poses a challenge for IO operations.

In GKE, Kubernetes tackles these constraints by housing necessary dependencies in a chroot environment (`/home/kubernetes/containerized_mounter/rootfs`) and mounting directories within it, enabling the execution of required tasks.

Longhorn provides a GKE COS node agent daemonset, which leverages GKE Kubernetes solutions to configure and run necessary dependencies. This agent is responsible for the following operations:

- Mounting the Longhorn data path.
- Loading the kernel module.
- Installing and running the iSCSI daemon.

## GKE COS Node Agent Installation

> 1. Configure Longhorn GKE COS node agent. You can also leave as default.
>    - Longhorn data paths: Updating the `node-agent` container's environment variable with a comma-separated list, for example:
>      ```yaml
>      containers:
>        - name: node-agent
>          env:
>            - name: LONGHORN_DATA_PATHS
>              value: /var/lib/longhorn1,/var/lib/longhorn2
>      ```

Reviewer 3: ```suggestion
1. Configure the Longhorn GKE COS node agent. You can use the default settings, if applicable.

    > **Tip:**
    > You can use a comma-separated list when specifying values for the `node-agent` container's environment variable (`LONGHORN_DATA_PATHS`).
    > 
    > Example:
    > 
    > ```yaml
    >  containers:
    >    - name: node-agent
    >      env:
    >        - name: LONGHORN_DATA_PATHS
    >          value: /var/lib/longhorn1,/var/lib/longhorn2
    > ```
```

1. Install the Longhorn GKE COS node agent.
    ```
    kubectl apply -f https://raw.githubusercontent.com/longhorn/longhorn/v{{< current-version >}}/deploy/prerequisite/longhorn-gke-cos-node-agent.yaml
    ```

> 1. Check the agent pod's status.

Reviewer 3: ```suggestion
1. Check the agent pod's status.

    Example:
```

    ```
    $ kubectl get pod -l app=longhorn-gke-cos-node
    NAME                                READY   STATUS    RESTARTS      AGE
    longhorn-gke-cos-node-agent-222w8   1/1     Running   1 (86m ago)   86m
    longhorn-gke-cos-node-agent-8r26h   1/1     Running   1 (86m ago)   86m
    longhorn-gke-cos-node-agent-nwhsw   1/1     Running   1 (86m ago)   86m
    ```

1. Check the installation result in the agent pod logs.
    ```
    Completed!
    Keep the container running for iscsi daemon
    ```
    > **Note:**

>     > The agent installs iSCSId within a container using package managers, hence you will likely to see error logs similar to `System has not been booted with systemd as init system (PID 1). Can't operate`. This occurs because the package manager tries to initiate iSCSI services through systemd. And, this causes issues because the container environment doesn't fully support systemd. To work around this, the script manually starts the daemon instead of relying on systemd. Therefore, in this context, such errors can be disregarded.

Reviewer 3: ```suggestion
    > The agent installs the iSCSI daemon (iscsid) in a container using a package manager. However, the package manager attempts to initiate iSCSI services through systemd, which the container environment does not fully support. As a result, you will likely see error logs similar to `System has not been booted with systemd as init system (PID 1). Can't operate`. To work around this, the script manually starts the daemon instead of relying on systemd. In this context, you can disregard the mentioned errors.
```

1. Verify that the dependent kernel module is loaded. You must run the command on the host.
    ```
    $ lsmod | grep -q iscsi_tcp && echo "The iSCSI module is loaded" || echo "The iSCSI module is NOT loaded"
    The iSCSI module is loaded
    ```

1. Verify that the iSCSI daemon is running. You must run the command on the host.
    ```
    $ ps aux | grep -q '[i]scsid' && echo "The iSCSI daemon is running" || echo "The iSCSI daemon is NOT running"
    The iSCSI daemon is running
    ```

> 1. Verify that the Longhorn data path (`/var/lib/longhorn`) is mounted on the host. If multiple Longhorn data paths are given, execute the command for each individual path. You must run the command on the host.

Reviewer 3: ```suggestion
1. Verify that the Longhorn data path (`/var/lib/longhorn`) is mounted on the host. If you specified multiple Longhorn data paths, run the command for each path on the host.
```
